### DISCUSSION TITLE
베이스라인 코드 설명

### DISCUSSION THREAD

**Discussion by rasauq1122** [boostcampaitech7/level2-nlp-generationfornlp-nlp-12-lv3#37] [Show and tell]
# 실행 방법
```
python src/main.py --config {config_name} --train # 훈련
python src/main.py --config {config_name} --inference # 추론
python src/main.py --config {config_name} --do_both # 훈련 후 바로 추론
```
- 실행 결과는 `experiment/{config_name}` 아래에 저장됩니다.
  - 생성된 `experiment/{config_name}/config.json`을 공유하여 실험의 상세를 팀원에게 알릴 수 있습니다.

# 왜 바꿨는가?
- 현재 베이스라인 코드는 `MyTrainer`, `MyInference` 두 클래스로 나누어져 있습니다. 각각 하는 일을 간략히 이야기하면 다음과 같습니다.
  - `MyTrainer`: csv 파일을 프롬프트 형태로 변환하여 Dataset 확보 후 모델 학습
  - `MyInference`: 학습한 모델을 불러와 모델이 생성한 logit을 기반으로 문제의 정답을 추출하여 저장
- 여기서 저는 logit 기반이 아닌 생성 기반으로 문제를 해결하고 싶어 위 클래스를 상속받아 생성 기반으로 문제를 해결하는 클래스를 만들고 싶었는데, 위와 같이 나누어져 있는 까닭에 하나의 일을 하는 두 가지 클래스를 만들어야 되는 상황이었습니다.
- 따라서 두 클래스를 하나로 합쳐 클래스의 응집도를 높이도록 하였습니다.

# 어떻게 바뀌었는가?
- 다음은 `config.json`으로부터 실험 설정을 불러와 베이스라인 모델을 학습시키는 코드입니다.
  ```python
  from pipeline import BasePipeline
  from manager import BaselineManager

  pipeline = BasePipeline("config", BaselineManager)
  pipeline.train()
  ```
- black-box 레벨에서 두 클래스에 대해 설명하면 다음과 같습니다.
  - `Pipeline`
    - csv 파일을 프롬프트 형태로 변환
    - 모델의 inference에 관여 
  - `Manager`
    - 모델, 토크나이저, 데이터 콜레이터를 불러와서 저장
    - 훈련시 트레이너의 설정값(ex. `LoRAConfig`, `SFTConfig`) 등에 관여
  - 구현 레벨에 대해서는 후술하겠습니다.

# 설계 의도
- 원래 하나의 클래스로 관리하려고 했는데, 모델 관련된 부분을 따로 분리하는 게 더 관리하기 쉬울 것 같아서 분리하였습니다.
  - 모델마다 데이터 콜레이터(혹은 `chat_template`)나 양자화 설정 등이 달라질 수 있도록
  - inference를 `manager` 클래스에 넣지 않고 `pipeline` 클래스에 넣어 놓은 이유는, 모델에 관련있다기 보다는 프롬프트를 어떻게 주냐에 더 관련있다고 생각해서 거기로 빼놨습니다.
    - `manager`에 넣으면 생성용 EXAONE, logit용 EXAONE이 따로따로 있어야 하는데 이건 좀 이상한 것 같아서 분리했습니다.
    
# 커스터마이즈
- 사실 docstring에 어느 정도 써놔서 무리없이 보실 수 있을 것 같긴 합니다만, 처음 보면 찾기 어려울 것 같아서 '이걸 바꾸고 싶으면 이 코드를 고쳐야 한다.' 정도로 알려드리겠습니다.
  - 프롬프트: 
    - `pipeline.make_user_message` : 사용자 메시지. ex. `"문제: {paragraph} ..."`
    - `pipeline.make_chat_message` : 시스템 메시지를 포함한 전체적인 채팅 구성. ex. `"지문을 읽고 질문의 답을 구하세요."`
      - 고급: `assistant` 메시지를 추가하려면 `_make_chat_message` 함수를 참고하세요.
   - 모델:
     - `manager.set_model` : 모델을 호출하는 부분. `self.model`에 할당해야 합니다.
     - `manager.set_data_collator` : `response_template` 등 추가처리가 필요할 경우
     - `manager.TEMPLATE` : `chat_template` 수정이 필요하다면 `__init__` 단계에서 처리해주어야 합니다.
   - 데이터셋:
     - `pipeline.load_dataset` : 추가적인 column이 있는 데이터셋을 불러올 때 (ex. reason column)

# Config 이해하기
- `config/` 디렉토리에 json 파일이 여러 개 있는데 여기서 실험의 다양한 부분을 통제할 수 있습니다.
- 아래는 조절할 수 있는 값들에 대한 내용입니다.
  - `params` : `LoRAConfig`, `SFTConfig`의 다양한 값을 조절할 수 있습니다. 
    - 너무 길어서 자세한 내용은 생략합니다.
  - `model` : `name` 필드에 모델의 이름 혹은 모델이 저장된 체크포인트 경로를 지정합니다.
  - `data` : 데이터셋을 어떻게 다룰지 조절할 수 있습니다.
    - `data_path` : 훈련용 파일, 추론용 파일 등이 어디에 저장되어있는지에 대한 정보입니다.
    - `train_file`, `test_file` : 훈련용 파일, 추론용 파일의 이름입니다.
    - `dev_file` : 검증용 데이터셋을 따로 넣어줄 생각이라면 사용하시면 됩니다.
    - `test_size` : `dev_file`이 null일때만 사용. 훈련용 데이터의 일부를 검증용 데이터셋으로 사용하고 싶을때 사용합니다.
    - `filtering_input_ids_length` : 데이터셋을 토큰으로 자른 뒤 길이가 너무 길면 필터링하고 싶을 때 사용합니다. ($-1$로 설정하여 아예 사용하지 않을 수 있습니다.)
    - `use_4-choices_prompt` : 4지선다라면 그에 맞는 프롬프트를 사용하고 싶을 때 사용합니다.
  - `experiment` : 실험의 전체적인 설정에 관여합니다.
    - `output_dir` : 실험 결과가 저장되는 파일 경로를 설정합니다.
    - `last_eval_strategy` : 훈련이 끝나고 evaluate를 어떻게 진행할지 정합니다.
      - `"no"` : 훈련이 끝나고 아무것도 안합니다.
      - `"evaluate"` : 검증용 데이터셋을 이용하여 지표(ex. 통계, loss 등)를 얻어 저장합니다.
      - `"inference"` : 검증용 데이터셋을 이용하여 추론을 진행합니다. 그 결과를 저장합니다.
    - `save_train_dataset` : 사용하고 있는 훈련용 데이터셋을 저장합니다.
    - `save_eval_dataset` : 사용하고 있는 검증용 데이터셋을 저장합니다.
## Config 커스터마이즈
- 새로운 config를 추가하려고 하여 사용하려고 할 때 다음과 같이 코드를 수정하면 됩니다.
- `pipeline`
  - `self.data_config.get("new-config", "default value")` 
  - `self.experiment_config.get("new-config", "default value")`
- `manager`
  - `self.model_config.get("new-config", "default value")`
  - `self.params.get("new-config", "default value")`
